### PR TITLE
[Diffusion] [NPU] Wan2.2-T2V-A14B-Diffusers modelslim quantization support

### PR DESCRIPTION
**Pull request by OrangeRedeng**
## Motivation

Support w8a8 (and w4a4) quantized [Wan2.2-I2V-A14B-Diffusers](https://modelers.cn/models/Wan-AI/Wan2.2-I2V-A14B-Diffusers) model by [modelslim](https://gitcode.com/Ascend/msmodelslim/blob/master/example/multimodal_sd/Wan2_1/README.md)

## Modifications

- [x] Add modelslim support for diffusion models (with full w4a4 dynamic, w8a8 static/dynamic linear support)
- [x] Modify wanvideo, fix prefix naming
- [x] Add wan_repack function to convert original wan from msmodelslim to diffusers format
- [x] Fix get model name for quantized models

## Accuracy Tests

To run quantized model in sglang, convert msmodelslim w8a8 model using:

`wan_repack.py  --input-path *quantized_model_path* --output-path *original Wan2.2-I2V-A14B-Diffusers model without transformer and transformer_2 folders"`

**Warning**: Sglang does not support quantized embeddings at the moment, let me know if this functionality is needed.

Then copy `config.json` from original transformer/transformer_2 folder to quantized transformer/transformer_2 folder.

Original:

https://github.com/user-attachments/assets/9b734aeb-121d-477c-a0cd-717ac56d0664

W8A8:

https://github.com/user-attachments/assets/4b6e3127-3a7f-4497-93db-c2716c3ec77c

## Benchmarking and Profiling

Command:
`SGLANG_CACHE_DIT_FN=2 SGLANG_CACHE_DIT_BN=1 SGLANG_CACHE_DIT_WARMUP=4 SGLANG_CACHE_DIT_RDT=0.4 SGLANG_CACHE_DIT_MC=4 SGLANG_CACHE_DIT_TAYLORSEER=true SGLANG_CACHE_DIT_TS_ORDER=2 SGLANG_CACHE_DIT_ENABLED=true  sglang generate --model-path *model_path* --prompt "Two anthropomorphic cats in comfy boxing gear and bright gloves fight intensely on a spotlighted stage." --height 720 --width 1280 --tp-size 4 --sp-degree 2 --num-gpus 8 --num-frames 81 --num-inference-steps 40`

w8a8 gives 7% acceleration compare to FP16 (325.47 vs 350.30)

## Checklist

- [x] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [x] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [x] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [x] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [x] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

## Review Process

1. Ping Merge Oncalls to start the PR flow. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process).
2. Get approvals from [CODEOWNERS](https://github.com/sgl-project/sglang/blob/main/.github/CODEOWNERS) and other reviewers.
3. Trigger CI tests with [comments](https://docs.sglang.io/developer_guide/contribution_guide.html#how-to-trigger-ci-tests) or contact authorized users to do so.
   - `/tag-run-ci-label`, `/rerun-failed-ci`, `/tag-and-rerun-ci`
4. After green CI and required approvals, ask Merge Oncalls to merge.

### CODE REVIEW

File: python/sglang/multimodal_gen/runtime/layers/quantization/modelslim.py [hunk 1]
```diff
@@ -0,0 +1,223 @@
...
+
+    def _get_scheme_from_parts(
+        self,
+        layer_name: str,
+    ) -> ModelSlimScheme:
+        
+        quant_type = self.quant_description.get(layer_name + ".weight", "")
+        print('+++++++++++++++++++++++')
+        print(layer_name)
+        print('----------------------+')
```
[the remark this answers is not in the document]
Comment (OrangeRedeng) [reply python/sglang/multimodal_gen/runtime/layers/quantization/modelslim.py]: removed

File: python/sglang/multimodal_gen/runtime/loader/fsdp_load.py [hunk 2]
```diff
@@ -232,7 +232,9 @@ def load_model_from_full_model_state_dict(
         full_sd_iterator, param_names_mapping
     )  # type: ignore
     for target_param_name, full_tensor in custom_param_sd.items():
+        print(target_param_name)
```
[the remark this answers is not in the document]
Comment (OrangeRedeng) [reply python/sglang/multimodal_gen/runtime/loader/fsdp_load.py]: removed

File: python/sglang/multimodal_gen/runtime/utils/hf_diffusers_utils.py [hunk 3]
```diff
@@ -321,6 +325,101 @@ def get_diffusers_component_config(
...
+    return quant_cfg
+
+
+def get_quant_config(
+    model_config,
+    packed_modules_mapping: Dict[str, List[str]] = {},
+    remap_prefix: Dict[str, str] | None = None,
+) -> QuantizationConfig:
+    print("*********************")
+    print(model_config["model_path"])
```
[the remark this answers is not in the document]
Comment (OrangeRedeng) [reply python/sglang/multimodal_gen/runtime/utils/hf_diffusers_utils.py]: removed

File: python/sglang/multimodal_gen/runtime/layers/quantization/modelslim.py [hunk 4]
```diff
@@ -0,0 +1,221 @@
...
+)
+from sglang.multimodal_gen.runtime.layers.quantization.base_config import (
+    QuantizationConfig,
+    QuantizeMethodBase,
+)
+
+if TYPE_CHECKING:
+    from sglang.srt.layers.quantization.modelslim.modelslim import ModelSlimConfig
+
+from types import MappingProxyType
```
Comment (ping1jing2) [inline python/sglang/multimodal_gen/runtime/layers/quantization/modelslim.py]: please move this import-statement into `if TYPE_CHECKING` block to reduce the time to import SGLang/SGL-Diffusion, and double check all the imports

File: python/sglang/multimodal_gen/runtime/layers/quantization/modelslim.py [hunk 5]
```diff
@@ -0,0 +1,221 @@
...
+    ModelSlimW4A4Int4,
+    ModelSlimW8A8Int8,
+)
+
+logger = logging.getLogger(__name__)
+
+
+class ModelSlimConfig(QuantizationConfig):
+    """
+    Config class for ModelSlim Quantization, a NPU-specific quantization type.
```
Comment (ping1jing2) [inline python/sglang/multimodal_gen/runtime/layers/quantization/modelslim.py]: please add more details about modelslim in docstring, at least we should add `python\sglang\srt\layers\quantization\modelslim\README.md`

File: python/sglang/multimodal_gen/runtime/layers/quantization/modelslim.py [hunk 6]
```diff
@@ -0,0 +1,221 @@
...
+    """
+
+    def __init__(self, quant_config: Dict[str, Any] = {}):
+        super().__init__()
+        self.quant_description = quant_config
+        ignore = cast(List[str], quant_config.get("ignore", []))
+        self.ignore = ignore if ignore is not None else []
```
Comment (ping1jing2) [inline python/sglang/multimodal_gen/runtime/layers/quantization/modelslim.py]: maybe we can also use `self.ignore = ignore or []`

File: python/sglang/multimodal_gen/runtime/layers/quantization/modelslim.py [hunk 7]
```diff
@@ -0,0 +1,221 @@
...
+        return cls(config)
+
+    def get_quant_method(
+        self,
+        layer: torch.nn.Module,
+        prefix: str,
+    ) -> Optional[QuantizeMethodBase]:
+        from sglang.multimodal_gen.runtime.layers.linear import LinearBase
+
+        if isinstance(layer, LinearBase):
```
Comment (ping1jing2) [inline python/sglang/multimodal_gen/runtime/layers/quantization/modelslim.py:84]: please use guard clause like below
```python
if not isinstance(layer, LinearBase):
    return None

# your normal logical code
```

File: python/sglang/multimodal_gen/runtime/layers/activation.py [hunk 8]
```diff
@@ -18,6 +18,12 @@
     from sgl_kernel import silu_and_mul
 # TODO (will): remove this dependency
 from sglang.multimodal_gen.runtime.layers.custom_op import CustomOp
+from sglang.multimodal_gen.runtime.platforms import current_platform
+
+if not current_platform.is_npu():
```
Comment (ping1jing2) [inline python/sglang/multimodal_gen/runtime/layers/activation.py]: it is hard to understand, please rewrite it like below
```python
if current_platform.is_npu():
    import torch_npu
else:
    from sgl_kernel import silu_and_mul
```

File: python/sglang/multimodal_gen/runtime/layers/layernorm.py [hunk 9]
```diff
@@ -33,6 +33,11 @@
 
 _is_cuda = current_platform.is_cuda()
 
+if not current_platform.is_npu():
```
Comment (ping1jing2) [inline python/sglang/multimodal_gen/runtime/layers/layernorm.py]: ditto

File: python/sglang/multimodal_gen/runtime/platforms/__init__.py [hunk 10]
```diff
@@ -101,6 +101,24 @@ def rocm_platform_plugin() -> str | None:
...
+    is_npu = False
+
+    try:
+        import torch
+
+        if torch.npu.is_available():
+            is_npu = True
+            logger.info("NPU is available")
+    except Exception as e:
+        logger.info("NPU detection failed: %s", e)
```
Comment (ping1jing2) [inline python/sglang/multimodal_gen/runtime/platforms/__init__.py:114]: maybe warning level is suitable

File: python/sglang/multimodal_gen/runtime/layers/quantization/modelslim.py (hunk 7 above)
Comment (OrangeRedeng) [reply python/sglang/multimodal_gen/runtime/layers/quantization/modelslim.py:84]: resolved

File: python/sglang/multimodal_gen/runtime/layers/quantization/modelslim.py (hunk 4 above)
Comment (OrangeRedeng) [reply python/sglang/multimodal_gen/runtime/layers/quantization/modelslim.py]: Thanks, resolved

Comment (OrangeRedeng) [reply python/sglang/multimodal_gen/runtime/layers/quantization/modelslim.py]: This does not work in this case

File: python/sglang/multimodal_gen/runtime/loader/fsdp_load.py [hunk 11]
```diff
@@ -142,7 +146,11 @@ def maybe_load_fsdp_model(
         if quant_method is not None and hasattr(
             quant_method, "process_weights_after_loading"
         ):
+            if _is_npu:
```
Comment (BBuf) [inline python/sglang/multimodal_gen/runtime/loader/fsdp_load.py:148]: Any comment can be added here?

File: python/sglang/multimodal_gen/runtime/models/dits/wanvideo.py [hunk 12]
```diff
@@ -149,16 +151,32 @@ def __init__(
 
         # layers
         self.to_q = ColumnParallelLinear(
```
Comment (BBuf) [inline python/sglang/multimodal_gen/runtime/models/dits/wanvideo.py:154]: Why we need `prefix` here?

File: python/sglang/multimodal_gen/runtime/loader/fsdp_load.py (hunk 11 above)
Comment (OrangeRedeng) [reply python/sglang/multimodal_gen/runtime/loader/fsdp_load.py:148]: Yes, of course, this is necessary to activate the NZ format for storing weights, which is a specific optimization for Ascend NPU. Add comment

File: python/sglang/multimodal_gen/runtime/models/dits/wanvideo.py (hunk 12 above)
Comment (OrangeRedeng) [reply python/sglang/multimodal_gen/runtime/models/dits/wanvideo.py:154]: Here (and elsewhere) a prefix declaration is needed for modelslim quantization to work. We have lines in the quant config like ```blocks.0.attn2.to_k.weight_scale": "W8A8_DYNAMIC"``` and we use prefixes to understand which quantization scheme to use for a particular layer.

File: python/sglang/multimodal_gen/runtime/utils/quantization_utils.py [hunk 13]
```diff
@@ -0,0 +1,165 @@
...
+
+    quant_config_str = metadata.get("_quantization_metadata")
+    if not quant_config_str:
+        return None
+    try:
+        quant_config_dict = json.loads(quant_config_str)
+    except Exception as _e:
+        return None
+
+    # handle diffusers fp8 safetensors metadata format
```
Comment (BBuf) [inline python/sglang/multimodal_gen/runtime/utils/quantization_utils.py:142]: The logic here is a bit hacky. Can we extract it into a separate function?

Comment (OrangeRedeng) [reply python/sglang/multimodal_gen/runtime/utils/quantization_utils.py:142]: Moving it into a separate function

Comment (OrangeRedeng) [reply python/sglang/multimodal_gen/runtime/utils/quantization_utils.py:142]: Is everything okay now? To be honest, this code was added in another PR https://github.com/sgl-project/sglang/pull/19142, I just moved it out as related to quantization in new quantization_utils.py file

**Review by BBuf — APPROVED** [review]
LGTM

**Review by ping1jing2 — APPROVED** [review]
